### PR TITLE
docs: add parameterized CMPs proposal

### PR DESCRIPTION
**Pull request by crenshaw-dev**

### CODE REVIEW

File: docs/proposals/cmp-parameters.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+repo, I would like to be able to pass values to the Helm chart without having to manually discover those parameter names
+(in other worse, they should show up in the Application UI just like with a native Helm Application). I also shouldn't 
+have to ask my Argo CD admin to modify the CMP to accommodate the values as environment variables.
+
+### Implementation Details/Notes/Constraints 
+
+#### Terms
+
+* **Parameter announcement**: an instance of a data structure which describes an individual parameter that may be applied
+  to a specific Application.
```
Comment (keithchong) [inline docs/proposals/cmp-parameters.md:115]: The current first class support for Helm in the UI has custom/specific interfaces, and they are different than, say, those for Kustomize. What were your thoughts on providing a general structure that will fit for all tools? Also, should we provide UI customizations that can be configured and contributed by each CMP, so they for example, can customize 

1. the ordering of their parameters (currently they are in alphabetical order)
2. custom groups with titles.  eg. instead of a.b.param1, a.b.param2 on different lines, why not show it as param1 and param2, under some logical grouping?
3. Allow different font styles for different parameters.  Eg. what if some parameters are required, so do you want the label to be bolded or underlined?
4. etc.

Basically, should we give the CMP owner the ability to customize the look and feel of the UI, based on some common configuration

Comment (crenshaw-dev) [reply docs/proposals/cmp-parameters.md:115]: Good questions. I'll answer here and also by adding info to the proposal. Some of the questions are answered in the notes I haven't transferred into the proposal yet. 😆 

> should we give the CMP owner the ability to customize the look and feel of the UI...

Yep! I'll add details to the "Easy to use" goal. But the basic idea is that parameter announcements should be able to communicate to the UI how they should be displayed.

> ...based on some common UI configuration spec that we define

We should define _some kind_ of schema. That schema will be on a spectrum from loosely- to strictly-defined.

On one end of the spectrum, we define a data structure that allows the user to describe the UI in excruciating detail. On the other end, we accept minimal information and let the UI make decisions from there.

Here's my suggestion: a parameter announcement should accept two strings, `type` and `description`. `type` tells the UI how to interpret `description`. If `type` is "string", the UI presents a simple test input box, and there's not really any need for further `description`. For other `type`s, there will be a documented schema for the `description` field. For example, if the type is `enum`, `description` might contain a JSON list of acceptable values.

The UI would implement components for common parameter types (like string, enum, multiline text, and image). CMP developers would have to contribute UI code for new types. We could offer a way for CMPs to inject UI code for their custom types, but I'd consider that out of scope for this proposal.

> the ordering of their parameters

The UI could respect the order of the list of parameter announcements from the CMP.

> custom groups with titles

Each parameter announcement could include a `section` field. The values could be used for grouping and for group headers.

> Allow different font styles for different parameters

That info could be included in the `description` field. To accommodate this, we'd probably want to define an initial `description` schema for the `text` type of "empty string or JSON object". That would be flexible enough to add features like formatting.

> what if some parameters are required

I'm not aware of any required parameters for the existing config management tools. But if we decided this was important, we could add a `required` field to the parameter announcement spec.

**Comment by jopit** [comment]
Replicating the behaviour of the Helm UI's `VALUE FILES` field will be challenging, since populating the drop-down  requires examining the contents of the git repo specified in the UI's application creation panel.  I don't see how the plugin itself could do that. Maybe we'd need something special in the `description` field to specify building the list by matching the files in the repo against a pattern.

image

**Comment by crenshaw-dev** [comment]
@jopit does the plugin not have access to the contents of the repo? Seems like it definitely has to have that access in order to build manifests.

File: docs/proposals/cmp-parameters.md (hunk 1 above)
Comment (crenshaw-dev) [reply docs/proposals/cmp-parameters.md:114]: @keithchong updated with some examples that should clarify.

Comment (keithchong) [reply docs/proposals/cmp-parameters.md:114]: > We should define some kind of schema. That schema will be on a spectrum from loosely- to strictly-defined.

Were you thinking this should be part of the plugin.yaml or a new file that the CMP needs to provide? I was trying to come up with the pros and cons. Should UI-specific details be separated out?

I do like the idea of having some core, base types, like what you described.  I know there is an issue with arrays not being shown properly in the UI but that could be improved upon.  (eg. show +/- to add or remove elements)

> That info could be included in the description field. 

I edited my comment after I posted it. We should also keep in mind to support translation for these user-friendly strings.

> I'm not aware of any required parameters for the existing config management tools. 

I not sure, but that could be something to consider in the general case. In the UI, if the field is left blank, it implies it is optional.  How will the user know which fields are required and need to be filled in?

Comment (crenshaw-dev) [reply docs/proposals/cmp-parameters.md:114]: > Were you thinking this should be part of the plugin.yaml or a new file that the CMP needs to provide?

The schema of the parameter announcement would be documented, and then the CMP would adhere to that schema when producing a JSON array of parameters for a given application. My parameter announcement schema proposal: https://github.com/argoproj/argo-cd/blob/3c972d4211b8d48d379d9b0fbede5329a5c3cb0f/docs/proposals/cmp-parameters.md#parameter-announcement-list-schema

> Should UI-specific details be separated out? 
> ...
> We should also keep in mind to support translation for these user-friendly strings.

I should rename the `description` field to `metadata` or something like that. I'm thinking of that field, not as a textual description of the parameter, but as a JSON object holding information about how the UI should present the parameter field.

> How will the user know which fields are required and need to be filled in?

It's a good question, but I think that's the status quo for native config management tools. One way this could be implemented is by adding UI support for this announcement:

```json
[
  {
    "name": "some-required-parameter",
    "description": "{\"required\": true}"
  }
]
```

**Comment by jopit** [comment]
@crenshaw-dev The plugin will have access at generation time (the repo is cloned into a shared volume), but I don't think it has access at the time the application is being created, which is when the prompt is needed.

**Comment by crenshaw-dev** [comment]
@jopit gotcha. Is there any reason it _couldn't_ have access to the repo at application creation time? Looks to me as if the native Helm config management code must be cloning the repo at application creation time to do things like populate the directory list and the parameters interface. Is there any reason the plugin couldn't take advantage of that just-cloned repo?

**Comment by jopit** [comment]
For the build-in helm, the parameters are different depending on the application, looking at the helm apps in https://github.com/argoproj/argocd-example-apps, the `helm-dependency` app has these parameters

image

but the `helm-guestbook` has these

image

**Comment by crenshaw-dev** [comment]
Parameter announcements could be requested/produced on a per-application basis though, right? That's the assumption I've been working with.

**Comment by jopit** [comment]
@crenshaw-dev I'm not actually sure how the built-in helm code gets those values.  I had assumed it wouldn't be cloning the repo on-the-fly when the user is filling in the app creation panel (before the app is created), because that seems potentially expensive and slow.  Could be wrong, though.  @alexmt could you shed some light on how this works?

**Comment by jopit** [comment]
Yeah, parameter announcements could/would be on per-application basis.  Although the UI needs the announcements before the app is actually created.  Which means the plugin is definitely going to need to be able to access the repo, and will have also have to be told the path to the manifests within that repo.

File: docs/proposals/001-proposal-template.md [hunk 2]
```diff
@@ -3,7 +3,7 @@ title: Neat-enhancement-idea
 authors:
   - "@sbose78" # Authors' github accounts here.
 sponsors:
-  - TBD        # List all intereste parties here.
```
Comment (alexmt) [inline docs/proposals/001-proposal-template.md:6]: 👍

File: docs/proposals/cmp-parameters.md [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+sense because 1) it's used for parameters announcements (consistency is good) and 2) JSON tooling is better.
+
+#### Parameter definition schema
+
+A parameter definition is an object with following schema:
+
+```go
+type ParameterDefinition struct {
+	// Name is the name of a parameter. (required)
+	Name string `json:"name"`
```
Comment (jopit) [inline docs/proposals/cmp-parameters.md:173]: Names should be unique, but are they unique only within the section, or should they be globally unique

Comment (crenshaw-dev) [reply docs/proposals/cmp-parameters.md:173]: I think they should be unique per-section. If my CMP has a parameter called, say, `address` that's applicable to all applications, I wouldn't want it to conflict if used in conjunction with a Helm chart that has a parameter of the same name.

**Comment by crenshaw-dev** [comment]
Marking this as ready for review since I think the bulk of the content is there. I'll add clarifications in response to comments. I'll probably also add screenshots of current parameters interfaces as well as mocked-up screenshots of the proposed parameters interfaces (hopefully with minimal differences).

File: docs/proposals/parameterized-config-management-plugins.md [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+metadata:
+  name: cmp-plugin
+spec:
+  version: v1.0
+  generate:
+    command: ["example.sh"]
+  discover:
+    fileName: "./subdir/s*.yaml"
+  # NEW KEY
+  parameters:
```
Comment (crenshaw-dev) [inline docs/proposals/parameterized-config-management-plugins.md:198]: It irks me that this isn't a verb. 😆

File: docs/proposals/parameterized-config-management-plugins.md [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+**get-parameters.sh**
+
+```shell
+# Pull params from values.yaml and then append the default ("main") parameters announcements.
+yq e -o=p values.yaml | jq -nR 'inputs | sub(" .*"; "") | {name: ., section: "Properties"}' |
+jq --slurp '[{"name": "values files", "uiConfig": "{\"multiline\": true}"}, {"name": "values"}] + .'
+```
+
+### Security Considerations
+
```
Comment (jessesuen) [inline docs/proposals/parameterized-config-management-plugins.md:444]: This section is noticeably absent. Since we are executing user-defined commands (operator-defined ones which are possibly affected by end-user input), we need to understand and document the risk here. A poorly written plugin could cause security issues (e.g. one that does not escape strings properly).

Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:466]: @jessesuen I structured the document a bit weirdly. That information is here, just one more heading down. :-)

If we add another `####` section under security, I'll keep the `Increased scripting` heading. Otherwise I'll drop it to avoid confusion.

File: docs/proposals/cmp-parameters.md (hunk 1 above)
Comment (keithchong) [reply docs/proposals/cmp-parameters.md]: Hi Michael, not sure if you saw in the chat during today's contributors call, but Jesse mentioned using a Open API to define the types.

One thing I was thinking about was the user friendliness/usability of the CMP UI. One thing I noticed was the current lack of (at least, from what I saw) validation of parameter values. Is the onus on the UI to perform validation on a specific type? Take for example, the Helm parameter `replicaCount`? If the user inadvertently enters a letter instead of a number, do we want up-front validation, or, we let the user proceed and let generation happen and then see the resulting failure (like it is now)?  How about semantic validation?  For semantic validation, it should be done by the CMP. Should there be `validate` (alongside `generate` and `discover`) and maybe validate should be done before generate.  Thoughts?

File: docs/proposals/parameterized-config-management-plugins.md [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+    3. Allow us to work around [this bug](https://github.com/argoproj/argo-cd/issues/7291) without including the Helm 
+       SDK in the core Argo CD code
+  * The Helm CMP must be on-par with the native implementation.
+    1. It must present an equivalent parameters UI.
+    2. It must communicate errors back to the repo-server (and then the UI) the same as the native implementation.
+
+### Non-Goals
+
+We should not:
+* Re-implement config management tools as CMPs (besides Helm)
```
Comment (leoluz) [inline docs/proposals/parameterized-config-management-plugins.md:152]: I thought we wanted implement Kustomize integration as CMP to dogfood the feature. Isn't it the case anymore?

File: docs/proposals/parameterized-config-management-plugins.md [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+
+As an Argo CD developer, I would like to be able to build Argo CD without including the Helm SDK as a dependency.
+
+The Helm SDK includes the Kubernetes code base. That's a lot of code, and it will make builds unacceptably slow.
+
+#### Use case 2: writing CMPs with rich UI experiences
+
+As an Argo CD user, I would like to be able to parameterize manifests built by a CMP.
+
+For example, if the Argo CD administrator has installed a CMP which applies a last-mile kustomize overlay to a Helm
```
Comment (leoluz) [inline docs/proposals/parameterized-config-management-plugins.md:168]: > which applies a last-mile kustomize overlay to a Helm repo

Not sure I understand this use-case. Does it mean that we are mixing Helm Charts and Kustomize in the same application?

File: docs/proposals/parameterized-config-management-plugins.md (hunk 6 above)
Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:152]: I could be wrong about this one. @alexmt to we want to do Kustomize-as-CMP for the initial release?

File: docs/proposals/parameterized-config-management-plugins.md (hunk 7 above)
Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:168]: @leoluz potentially. The specific use case I'm thinking of is when someone wants to use an external Helm chart (say, minio), but then they want to add a secret or apply an annotation or some other minor change.

This use case could be solved by creating an internal repo (https://github.company.com/org/kustomized-minio.git) which pulls in the minio Helm chart (using the kustomize `helmCharts` feature) but then applies a kustomize overlay.

You can do that with Argo CD today. But then you won't be able to set any of the minio Helm parameters in the Argo CD UI.

A CMP with robust parameter discovery could 1) parse the kustomization.yaml to find all the `helmCharts`, 2) perform parameter discovery on each Helm chart, and 3) announce each Helm chart's parameters in their own "section" in the parameter announcement.

File: docs/proposals/parameterized-config-management-plugins.md [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+  version: v1.0
+  generate:
+    command: ["example.sh"]
+  discover:
+    fileName: "./subdir/s*.yaml"
+  # NEW KEY
+  parameters:
+    # The declarative announcement follows the parameters announcement schema. This is where a parameter description
+    # should go if it applies to all apps for this CMP.
+    announcement:
```
Comment (leoluz) [inline docs/proposals/parameterized-config-management-plugins.md:214]: It seems to me that announcement is more an implementation concern and Im not sure if it should be part of this contract. 

Why not having something simpler like:

```yaml
  parameters:
  - name: example
  dynamicParameters:
  - command: ["example-params.sh"]
```

In this case both would be merged and announced to the UI.

Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:214]: Yep, that's a lot simpler. Will change the proposal.

File: docs/proposals/cmp-parameters.md (hunk 1 above)
Comment (crenshaw-dev) [reply docs/proposals/cmp-parameters.md]: I think validation is definitely something we should be ready to support with parameterized CMPs. I'm not sure what role Open API (or maybe JSON schema, if that's different?) might play, but certainly worth looking at.

I'm not sure we should define a `validate` field yet, since that would be adding a feature that doesn't exist for config management tools. But I could imagine validation info being part of the `uiConfig` field (renamed from `metadata` field described above). Putting that info in the `uiConfig` field for now would leave room for a short-term improvement until we're ready to fully-define a validation system.

Can you give an example of what you mean by semantic validation?

File: docs/proposals/parameterized-config-management-plugins.md [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+
+## Open Questions
+
+Is there a way to make it easier for the generate command to consume parameters? (Easier than parsing a JSON object, 
+that is.)
+* We could translate all parameters to env vars, like `ARGOCD_PARAM_{section}_{param name}`.
+  * Advantage: no need to pull the param from JSON.
+  * Disadvantage: we'd need an easy-to-predict and collision-resistant way to translate arbitrary parameter names
+    to env vars. 
+    * Option 1: Enforce some parameter name convention which allows a 1:1 translation to env
```
Comment (jopit) [inline docs/proposals/parameterized-config-management-plugins.md:70]: For option 1, the plugin might not have control over the parameter names.  For example, the parameters for the helm built-in come from whatever the application creator puts in the values file.

File: docs/proposals/parameterized-config-management-plugins.md [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+      vars. Is limiting the param namespace worth it for the usability win?
+    * Option 2: Only translate to an env var if the param name happens to be in the safely-translatable
+      namespace. For example, `set-file` becomes `SET_FILE`, but `omg!!!param` doesn't get translated to an env var.
+    * Option 3: Add an `envVar` field to the [parameter definition schema](#parameter-definition-schema) to specify
```
Comment (jopit) [inline docs/proposals/parameterized-config-management-plugins.md:74]: If the plugin doesn't control the parameter names, it might have a hard time specifying the `envVar` field.

File: docs/proposals/parameterized-config-management-plugins.md [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+
+Is there a way to make it easier for the generate command to consume parameters? (Easier than parsing a JSON object, 
+that is.)
+* We could translate all parameters to env vars, like `ARGOCD_PARAM_{section}_{param name}`.
+  * Advantage: no need to pull the param from JSON.
+  * Disadvantage: we'd need an easy-to-predict and collision-resistant way to translate arbitrary parameter names
+    to env vars. 
+    * Option 1: Enforce some parameter name convention which allows a 1:1 translation to env
+      vars. Is limiting the param namespace worth it for the usability win?
+    * Option 2: Only translate to an env var if the param name happens to be in the safely-translatable
```
Comment (jopit) [inline docs/proposals/parameterized-config-management-plugins.md:72]: Is the assumption here that the JSON `ARGOCD_PARAMETERS` env var will also be set, so that the `omg!!!param` can still be found?

File: docs/proposals/parameterized-config-management-plugins.md [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+```go
+type ParameterDefinition struct {
+	// Name is the name of a parameter. (required)
+	Name string `json:"name"`
+	// Type is the type of the parameter. This determines the schema of `uiConfig` and how the UI presents the 
+	// parameter. (default is `string)
+	Type string `json:"type"`
+	// UiConfig is a stringified JSON object containing information about how the UI should present the parameter. 
+	// (default is `{}`)
+	UiConfig string `json:"uiConfig"`
```
Comment (jopit) [inline docs/proposals/parameterized-config-management-plugins.md:274]: I realize we want to keep the `uiConfig` open ended, but shouldn't we come up with an initial set of supported interface descriptions for it?  It feels to me that it's more than just an implementation detail.

File: docs/proposals/parameterized-config-management-plugins.md (hunk 11 above)
Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:72]: @jopit yep! `ARGOCD_PARAMETERS` would be guaranteed to always have all parameters.

File: docs/proposals/parameterized-config-management-plugins.md (hunk 10 above)
Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:74]: @jopit agreed. This would basically kick the can down the road to the plugin author. I guess one could argue that the plugin author might have more knowledge of the parameter names space and be able to write a better name -> envVar algorithm than we could.

File: docs/proposals/parameterized-config-management-plugins.md (hunk 12 above)
Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:274]: @jopit maybe... If there are any already-existing UI features we need to support, then I'd recommend defining that part of the `uiConfig` spec now. I haven't seen any features that wouldn't already be covered by `type`, but I'll take another look.

Aside: do you think `UiConfig` should be a generic JSON object instead of a stringified JSON object? If we know we want JSON, I'm not sure there's a reason to require it be stringified.

File: docs/proposals/parameterized-config-management-plugins.md [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+```
+
+The currently-configured parameters (if there are any) will be communicated to both `generate.command` and 
+`parameters.command` via an `ARGOCD_PARAMETERS` environment variable. The parameters will be encoded according to the 
+[parameters serialization format](#parameters-manifest--parameters-serialization-format) defined below.
+
+Passing the parameters to the `parameters.command` will allow configuration of parameter discovery. For example:
+
+```yaml
+plugin:
```
Comment (leoluz) [inline docs/proposals/parameterized-config-management-plugins.md:227]: It seems that this is the `Application` resource. If so, I'm not sure it is a good idea to manipulate fields in the `spec` section as it could potentially generate conflicts with the desired state (git). Maybe it is better to move the `parameters` attribute into the `status` section. 

Something like:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  ...
spec: 
  ...
status:
  cmp:
    last-applied-parameters:
      main:
      - name: ignore-helm-charts
        value: '["chart-a", "chart-b"]'
```

File: docs/proposals/parameterized-config-management-plugins.md (hunk 12 above)
Comment (leoluz) [reply docs/proposals/parameterized-config-management-plugins.md:274]: Just to clarify things, my suggestion was to maybe use [json.RawMessage](https://pkg.go.dev/encoding/json#RawMessage) instead of `string`.

Comment (jopit) [reply docs/proposals/parameterized-config-management-plugins.md:274]: I don't know of a reason it needs to be stringified.

The example down below shows the `enum` type needs a `uiConfig`

`types` like the following probably don't need `uiConfig`
- boolean
- string
- password

Would we want to define types and corresponding uiConfigs for things like the following?
- something like an enum, but it lets you select more than one from a list (build-in helm values file uses this), not sure what a good name for it is
- as above, but also lets you specify a value not on the list

File: docs/proposals/cmp-parameters.md (hunk 1 above)
Comment (keithchong) [reply docs/proposals/cmp-parameters.md]: I think the Open API definition can be used to verify the validity of `uiConfig`, when the plugin author is customizing the look of the UI, at “development time”, in other words. It should be versioned.

For semantic validation, perhaps take the Helm example, where the user has defined a values.yaml file in the repo, and the parameter value for the path is correct. It’s fine syntactically. However, semantically, it’s wrong because the file is not in the default path that Helm expects.  The UI can’t validate this semantically and it’s very specific to the plugin tool itself.

**Review by alexmt — APPROVED** [review]
LGTM.

Disclaimer: me, @crenshaw-dev , and @leoluz interact every day, so we've discussed the proposed design offline and agreed on the design. So approving without asking additional questions :)

File: docs/proposals/parameterized-config-management-plugins.md [hunk 14]
```diff
@@ -0,0 +1,840 @@
...
+echo "$VALUES" > /tmp/values.yaml
+helm template --values /tmp/values.yaml .
+```
+
+Environment variable names are set according to these rules:
+
+1. If a parameter is a `string`, the format is `escaped({name})` (`escaped` is defined below).
+2. If a parameter is an `array`, the format is `escaped({name}_{index})` (where the first index is 0).
+3. If a parameter is a `map`, the format is `escaped({name}_{key})`.
+4. If an escaped env var name matches one in the [build environment](https://argo-cd-docs.readthedocs.io/en/latest/user-guide/build-environment/),
```
Comment (jopit) [inline docs/proposals/parameterized-config-management-plugins.md:300]: An escaped env var could also match one of the shell's env vars, like `PATH`.  Is that a concern from a security point of view?

Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:300]: Excellent catch. Uncommenting the line below does cause the command to fail:

```go
func main() {
	cmd := exec.Command("sh", "-c", "ls")
	//cmd.Env = []string{"PATH="}
	out, err := cmd.Output()
	if err != nil {
		panic(fmt.Sprintf("failed on output: %s", err))
	}
	fmt.Println(string(out))
}
```

What would you think about prefixing all env vars with `PARAM_`?

Comment (jopit) [reply docs/proposals/parameterized-config-management-plugins.md:300]: That's probably the easiest way to prevent the problem.

Comment (crenshaw-dev) [reply docs/proposals/parameterized-config-management-plugins.md:305]: Updated. Thanks again for catching that!
